content/oci/oci.go

```diff
@@ -143,11 +160,55 @@ func (s *Store) Exists(ctx context.Context, target ocispec.Descriptor) (bool, er
 
 // Delete deletes the content matching the descriptor from the store. Delete may
 // fail on certain systems (i.e. NTFS), if there is a process (i.e. an unclosed
-// Reader) using target.
+// Reader) using target. If s.AutoGC is set to true, Delete will recursively
+// remove the dangling nodes caused by the current delete. If s.AutoRemoveReferrers
+// is set to true, Delete will recursively remove the referrers of the manifests
+// being deleted.
 func (s *Store) Delete(ctx context.Context, target ocispec.Descriptor) error {
-	s.sync.Lock()
-	defer s.sync.Unlock()
+	deleteQueue := []ocispec.Descriptor{target}
+
+	for len(deleteQueue) > 0 {
+		head := deleteQueue[0]
+		deleteQueue = deleteQueue[1:]
+
+		// get referrers if applicable
+		if s.AutoDeleteReferrers && descriptor.IsManifest(head) {
+			referrers, err := registry.Referrers(ctx, s, head, "")
+			if err != nil {
+				return err
+			}
+			deleteQueue = append(deleteQueue, referrers...)
+		}
+
+		// delete the head of queue
+		if err := func() error {
+			s.sync.Lock()
+			defer s.sync.Unlock()
```

Comment:
Continue the thread of https://github.com/oras-project/oras-go/pull/653#discussion_r1437397385 here

Can you lock/unlock at the beginning of the method? Otherwise, graphs may be changed when deleting.

The `Referrers()` method calling `Predecessors()` is indeed a problem. However, we could have a way to bypass it. Consider,
```go
type unsafeStore {
    *Store
}

func (s *Store) Predecessors(ctx context.Context, node ocispec.Descriptor) ([]ocispec.Descriptor, error) {
    return s.graph.Predecessors(ctx, node)
}

// then in Delete(), write this to bypass the lock
referrers, err := registry.Referrers(ctx, &unsafeStore{s}, head, "")
```

Reply:
change applied.